design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,189 @@
+ a simple event sourcing / CQRS style application.
+ We should provide a new ksqlDB client that provides that currently missing single interaction point with the event streaming
+ platform so that we can improve the application development process.
+           
+## Motivation and background
+
+In order to increase adoption of ksqlDB we need the developer experience to be as slick as possible. Currently, in order to write
+a simple event streaming application with ksqlDB, up to 4 clients may be needed: 
+
+1. An HTTP client to access the current HTTP/REST API
+2. A WebSocket client to access the streaming query WebSockets endpoint
+3. A JDBC client to access data in JDBC tables that are synced with KSQL via connect, as the current support for
+pull queries does not make the data otherwise easily accessible.
+4. A Kafka client for producing and consuming messages from topics.
+
+This is a lot for an application to handle and creates a steep learning curve for application developers.
+
+Moreover, our current HTTP API is lacking, especially in the area of streaming query results which makes it difficult to write an application
+that uses this API.
+
+This KLIP proposes that we:
+
+* Create a ksqlDB client (initially in Java and JavaScript) as the primary interaction point to the event streaming platform
+for an application
+* Create a new server side implementation that supports the client and enables the functionality of the client to be delivered in a straightforward
+and efficient way
+* The creation of a new wire protocol for the client<->server communication.
+* Migrate the functionality of the current HTTP/REST API over to the new server implementation and retire the parts that are no longer needed.
+
+## What is in scope
+
+### The client
+
+* We will create a new client, initially in Java (and potentially in JavaScript). Clients for other languages such as Python and Go will follow later.
+* The client will initially support execution and streaming of queries (both push and pull), inserting of rows into streams, DDL operations and admin operations
+such as list and describe.
+* The client will support a reactive / async interaction style using Java CompletableFuture and Reactive Streams / JDK9 Flow API (or a copy thereof if we cannot upgrade from Java 8)
+* The client will also support a direct / synchronous interaction style.
+
+### The protocol
+
+We will create a new super simple streaming protocol designed to efficiently stream data from server to client or client to server.
+HTTP/REST is a poor fit for streaming data as it is designed for a request/response interaction style. Attempting to implement "streaming"
+over HTTP/REST usually results in a polling approach which is awkward and inefficient compared to a protocol which is designed with streaming in mind from
+the start.
+
+* Binary multiplexed protocol. The protocol will support multiple independent channels over a single connection. We want to discourage clients from opening new
```

Comment:
Again this feels like re-inventing something thats in http/2 already? Could we reuse that ? It may be much easier for others in community to write new language clients that way?

Reply:
Please see previous comment about HTTP 2

Reply:
> Again this feels like re-inventing something thats in http/2 already? 

Binary multiplexed streaming protocols existed long before HTTP 2, so if anything it's HTTP 2 doing the "re-inventing" ;) Not that that is a bad thing, it's the best way to design a streaming protocol and that's why just about every messaging system which dominated the market before Kafka came along does something similar. (Kafka itself is notably the odd one out in not using a streaming protocol but taking a batching and RPC approach).